custom_components/localtuya/coordinator.py

```diff
@@ -380,39 +383,36 @@ async def update_local_key(self):
         await cloud_api.async_get_devices_list(force_update=True)
 
         cloud_devs = cloud_api.device_list
-        if not dev_id in cloud_devs:
-            return # Don't delete or disable such a device: it may be back later!
-        if not (cloud_localkey := cloud_devs[dev_id].get(CONF_LOCAL_KEY)):
-            return # TODO: May it happen? What does "no local key" for a device mean?
-        if self._local_key == cloud_localkey:
-            return
+        if dev_id in cloud_devs:
+            cloud_localkey = cloud_devs[dev_id].get(CONF_LOCAL_KEY)
+            if not cloud_localkey or self._local_key == cloud_localkey:
+                return
 
-        self._local_key = cloud_localkey
-        new_data = self._entry.data.copy()
+            new_data = self._entry.data.copy()
+            self._local_key = cloud_localkey
 
-        if self._node_id:
-            from .core.helpers import get_gateway_by_deviceid
+            if self._node_id:
+                from .core.helpers import get_gateway_by_deviceid
 
-            # Update Node ID.
-            if (new_node_id := cloud_devs[dev_id].get(CONF_NODE_ID)) and new_node_id != self._node_id:
-                new_data[CONF_DEVICES][dev_id][CONF_NODE_ID] = new_node_id
+                # Update Node ID.
+                if new_node_id := cloud_devs[dev_id].get(CONF_NODE_ID):
+                    new_data[CONF_DEVICES][dev_id][CONF_NODE_ID] = new_node_id
 
-            # Update Gateway ID and IP
-            if new_gw := get_gateway_by_deviceid(dev_id, cloud_devs):
-                self.info(f"Gateway ID has been updated to: {new_gw.id}")
-                new_data[CONF_DEVICES][dev_id][CONF_GATEWAY_ID] = new_gw.id
+                # Update Gateway ID and IP
+                if new_gw := get_gateway_by_deviceid(dev_id, cloud_devs):
+                    self.info(f"Gateway ID has been updated to: {new_gw.id}")
+                    new_data[CONF_DEVICES][dev_id][CONF_GATEWAY_ID] = new_gw.id
 
-                discovery = self._hass.data[DOMAIN].get(DATA_DISCOVERY)
-                if discovery and (local_gw := discovery.devices.get(new_gw.id)):
-                    new_ip = local_gw.get(CONF_TUYA_IP, self._device_config.host)
-                    new_data[CONF_DEVICES][dev_id][CONF_HOST] = new_ip
-                    self.info(f"IP has been updated to: {new_ip}")
+                    discovery = self._hass.data[DOMAIN].get(DATA_DISCOVERY)
+                    if discovery and (local_gw := discovery.devices.get(new_gw.id)):
+                        new_ip = local_gw.get(CONF_TUYA_IP, self._device_config.host)
+                        new_data[CONF_DEVICES][dev_id][CONF_HOST] = new_ip
+                        self.info(f"IP has been updated to: {new_ip}")
 
-        new_data[CONF_DEVICES][dev_id][CONF_LOCAL_KEY] = self._local_key
-        new_data[ATTR_UPDATED_AT] = str(int(time.time() * 1000))
-        # This will cause restart of LocalTuya
```

Comment:
I think, this comment is important! It is not obvious, that call to `async_update_entry` causes restart, and that's why there is no a need to update `self._gateway`, `self._node_id` and other `self` members here. In fact, updating `self._local_key` is not required here for the same reason.

Reply:
Yeah I know that it useless to update private variables if localkey changed I just let it be since upstream, 

I actually I don't think the comment is matter because it mentioned in `async_update_entry` docstring

![image](https://github.com/user-attachments/assets/1f56d893-727f-498d-a137-a46dfd292239)

Reply:
1. It is not obvious that `update_listeners` restarts LocalTuya.
2. I never saw such descriptions. I use mostly Studio Code Server HA plugin for editing, and Github for reading.

AFAIK, even in offline tools designed for Python programming, special efforts are required to see that descroption, while the comment is always in place.

Reply:
> It is not obvious that update_listeners restarts LocalTuya

It does, Again if you read the docstrings.

> I never saw such descriptions. I use mostly Studio Code Server HA plugin for editing, and Github for reading

^ Setup development environment. 

`async_update_entry` actually fires update_listeners if top-level data changed, basically this will fires the function in `__init__.py`  

https://github.com/xZetsubou/hass-localtuya/blob/dc04c5b33687649192d9968a0b29d9a152eb12af/custom_components/localtuya/__init__.py#L415-L417